enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+---
+title: control-plane-machineset-vertical-scaling
+authors:
+  - "@bergmannf"
+reviewers:
+  - @"JoelSpeed"
+approvers:
+  - @"JoelSpeed"
+api-approvers: 
+  - None
+creation-date: 2023-10-30
+last-updated: 2023-10-30
+tracking-link:
+  - https://issues.redhat.com/browse/OSD-15261
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Control plane machineset automatic vertical scaling
+
+## Summary
+
+To allow control plane nodes to remain at an adequate size for its cluster, this
+proposal introduces new configuration for the control plane machineset operator
+to allow it to make automated scaling decisions for control plane node sizing.
+
+## Motivation
+
+During normal operation of an OpenShift cluster, as worker node count and
+workloads increase, it becomes necessary to increase resources available to the
+control plane nodes.
+
+During SRE operations of the managed OpenShift product, there was already a
+trigger indentified, when these increases are required. Instead of requiring the
+adjustments to be performed by hand, the operator should be able to
+automatically determine if a control plane node size change is required.
+
+On the other hand, it should also be possible to determine that control plane
+nodes have become too large for a cluster, because load has decreased. In this
+case the operator should also be able to automatically decide to scale the
+control plane down again.
+
+With the automatic increase of the control plane size, cluster performance can
```

Comment:
Have we considered how to avoid flapping? So that we aren't in a constant loop of scale up/down as the capacity to usage ratios change during a rollout?

Reply:
The idea is to prevent hasty scale up and down using the two configuration variables `syncPeriod` and `gracePeriod`.
SyncPeriod is the minimum amount of time the average load threshold must be over oder under the scaleup and down threshold, before an actual change is performed.

Reply:
@elmiko Can you add context on how the Cluster Autoscaler adds hysteresis, I think it would be good to align